File: tools4a-args/src/androidTest/java/com/github/panpf/tools4a/args/test/read/TestBothIntentUriActivity.java
/*
 * Copyright (C) 2020 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.panpf.tools4a.args.test.read;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;

import com.github.panpf.tools4a.args.Argsx;

import org.junit.Assert;

public class TestBothIntentUriActivity extends FragmentActivity {

    @NonNull
    public static Intent createIntentWithUriAndExtras(@NonNull Context context) {
        String params = "byteIntentUriOrDefault=" + ((byte) 1) +
                "&shortIntentUriOrDefault=" + ((short) 2) +
                "&intIntentUriOrDefault=" + 3 +
                "&longIntentUriOrDefault=" + 4L +
                "&floatIntentUriOrDefault=" + ((float) 5) +
                "&doubleIntentUriOrDefault=" + ((double) 6) +
                "&booleanIntentUriOrDefault=" + true +
                "&stringIntentUriRequired=" + "stringIntentUriRequired" +
                "&stringIntentUriOptional=" + "stringIntentUriOptional" +
                "&stringIntentUriOrDefault=" + "stringIntentUriOrDefault" +

                //Activity Uri Intent
                "&byteUriIntentOrDefault=" + ((byte) 11) +
                "&shortUriIntentOrDefault=" + ((short) 12) +
                "&intUriIntentOrDefault=" + 13 +
                "&longUriIntentOrDefault=" + 14L +
                "&floatUriIntentOrDefault=" + ((float) 15) +
                "&doubleUriIntentOrDefault=" + ((double) 16) +
                "&booleanUriIntentOrDefault=" + true +
                "&stringUriIntentRequired=" + "stringUriIntentRequired" +
                "&stringUriIntentOptional=" + "stringUriIntentOptional" +
                "&stringUriIntentOrDefault=" + "stringUriIntentOrDefault";

        Uri uri = Uri.parse("tools4a://tools4a.com/panpf/tools4a/uri/intent?" + params);

        Intent starter = new Intent(context, TestBothIntentUriActivity.class);
        starter.setAction(Intent.ACTION_VIEW);
        starter.setData(uri);
        starter.putExtra("byteIntentUriOrDefault", ((byte) -1));
        starter.putExtra("shortIntentUriOrDefault", ((short) -2));
        starter.putExtra("intIntentUriOrDefault", -3);
        starter.putExtra("longIntentUriOrDefault", -4L);
        starter.putExtra("floatIntentUriOrDefault", -5f);
        starter.putExtra("doubleIntentUriOrDefault", (-6d));
        starter.putExtra("booleanIntentUriOrDefault", true);
        starter.putExtra("stringIntentUriRequired", "stringIntentRequired");
        starter.putExtra("stringIntentUriOptional", "stringIntentOptional");
        starter.putExtra("stringIntentUriOrDefault", "stringIntentOrDefault");

        //Activity Uri Intent
        starter.putExtra("byteUriIntentOrDefault", ((byte) -11));
        starter.putExtra("shortUriIntentOrDefault", ((short) -12));
        starter.putExtra("intUriIntentOrDefault", -13);
        starter.putExtra("longUriIntentOrDefault", -14L);
        starter.putExtra("floatUriIntentOrDefault", -15f);
        starter.putExtra("doubleUriIntentOrDefault", (-16d));
        starter.putExtra("booleanUriIntentOrDefault", true);
        starter.putExtra("stringUriIntentRequired", "stringUriIntentRequired");
        starter.putExtra("stringUriIntentOptional", "stringUriIntentOptional");
        starter.putExtra("stringUriIntentOrDefault", "stringUriIntentOrDefault");
        return starter;
    }

    public void checkParams() {
        Activity activity = this;

        byte byteIntentUriOrDefault = Argsx.readByteIntentUriArgOr(activity, "byteIntentUriOrDefault", ((byte) 0));
        short shortIntentUriOrDefault = Argsx.readShortIntentUriArgOr(activity, "shortIntentUriOrDefault", ((short) 0));
        int intIntentUriOrDefault = Argsx.readIntIntentUriArgOr(activity, "intIntentUriOrDefault", 0);
        long longIntentUriOrDefault = Argsx.readLongIntentUriArgOr(activity, "longIntentUriOrDefault", 0L);
        float floatIntentUriOrDefault = Argsx.readFloatIntentUriArgOr(activity, "floatIntentUriOrDefault", ((float) 0));
        double doubleIntentUriOrDefault = Argsx.readDoubleIntentUriArgOr(activity, "doubleIntentUriOrDefault", ((double) 0));
        boolean booleanIntentUriOrDefault = Argsx.readBooleanIntentUriArgOr(activity, "booleanIntentUriOrDefault", false);
        String stringIntentUriRequired = Argsx.readStringIntentUriArgOrThrow(activity, "stringIntentUriRequired");
        String stringIntentUriOrDefault = Argsx.readStringIntentUriArgOr(activity, "stringIntentUriOrDefault", "default");
        String stringIntentUriOrDefaultErrKey = Argsx.readStringIntentUriArgOr(activity, "stringIntentUriOrDefaultErrKey", "stringIntentUriOrDefaultErrKey");
        String stringIntentUriOptional = Argsx.readStringIntentUriArgOrNull(activity, "stringIntentUriOptional");
        String stringIntentUriOptionalErrKey = Argsx.readStringIntentUriArgOrNull(activity, "stringIntentUriOptionalErrKey");

        //uri intent
        byte byteUriIntentOrDefault = Argsx.readByteUriIntentArgOr(activity, "byteUriIntentOrDefault", ((byte) 0));
        short shortUriIntentOrDefault = Argsx.readShortUriIntentArgOr(activity, "shortUriIntentOrDefault", ((short) 0));
        int intUriIntentOrDefault = Argsx.readIntUriIntentArgOr(activity, "intUriIntentOrDefault", 0);
        long longUriIntentOrDefault = Argsx.readLongUriIntentArgOr(activity, "longUriIntentOrDefault", 0L);
        float floatUriIntentOrDefault = Argsx.readFloatUriIntentArgOr(activity, "floatUriIntentOrDefault", ((float) 0));
        double doubleUriIntentOrDefault = Argsx.readDoubleUriIntentArgOr(activity, "doubleUriIntentOrDefault", ((double) 0));
        boolean booleanUriIntentOrDefault = Argsx.readBooleanUriIntentArgOr(activity, "booleanUriIntentOrDefault", false);
        String stringUriIntentRequired = Argsx.readStringUriIntentArgOrThrow(activity, "stringUriIntentRequired");
        String stringUriIntentOrDefault = Argsx.readStringUriIntentArgOr(activity, "stringUriIntentOrDefault", "default");
        String stringUriIntentOrDefaultErrKey = Argsx.readStringUriIntentArgOr(activity, "stringUriIntentOrDefaultErrKey", "stringUriIntentOrDefaultErrKey");
        String stringUriIntentOptional = Argsx.readStringUriIntentArgOrNull(activity, "stringUriIntentOptional");
        String stringUriIntentOptionalErrKey = Argsx.readStringUriIntentArgOrNull(activity, "stringUriIntentOptionalErrKey");

        Assert.assertTrue(byteIntentUriOrDefault == ((byte) -1));
        Assert.assertTrue(shortIntentUriOrDefault == ((short) -2));
        Assert.assertTrue(intIntentUriOrDefault == -3);
        Assert.assertTrue(longIntentUriOrDefault == -4L);
        Assert.assertTrue(floatIntentUriOrDefault == -5f);
        Assert.assertTrue(doubleIntentUriOrDefault == -6d);
        Assert.assertTrue(booleanIntentUriOrDefault);
        Assert.assertTrue(stringIntentUriRequired.equals("stringIntentRequired"));
        Assert.assertTrue(stringIntentUriOptional.equals("stringIntentOptional"));
        Assert.assertNull(stringIntentUriOptionalErrKey);
        Assert.assertTrue(stringIntentUriOrDefault.equals("stringIntentOrDefault"));
        Assert.assertTrue(stringIntentUriOrDefaultErrKey.equals("stringIntentUriOrDefaultErrKey"));

        //Activity Uri Intent
        Assert.assertTrue(byteUriIntentOrDefault == ((byte) 11));
        Assert.assertTrue(shortUriIntentOrDefault == ((short) 12));
        Assert.assertTrue(intUriIntentOrDefault == 13);
        Assert.assertTrue(longUriIntentOrDefault == 14L);
        Assert.assertTrue(floatUriIntentOrDefault == 15f);
        Assert.assertTrue(doubleUriIntentOrDefault == 16d);
        Assert.assertTrue(booleanUriIntentOrDefault);
        Assert.assertTrue(stringUriIntentRequired.equals("stringUriIntentRequired"));
        Assert.assertTrue(stringUriIntentOptional.equals("stringUriIntentOptional"));
        Assert.assertNull(stringUriIntentOptionalErrKey);
        Assert.assertTrue(stringUriIntentOrDefault.equals("stringUriIntentOrDefault"));
        Assert.assertTrue(stringUriIntentOrDefaultErrKey.equals("stringUriIntentOrDefaultErrKey"));
    }
}

File: tools4a-args-ktx/src/androidTest/java/com/github/panpf/tools4a/args/ktx/test/read/TestBothIntentUriActivity.kt
/*
 * Copyright (C) 2020 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.panpf.tools4a.args.ktx.test.read

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.github.panpf.tools4a.args.ktx.*
import org.junit.Assert

class TestBothIntentUriActivity : androidx.fragment.app.FragmentActivity() {

    companion object {
        fun createIntentWithUriAndExtras(context: Context): Intent {
            val params = StringBuilder()
                    .append("byteIntentUriOrDefault=").append(1.toByte().toInt())
                    .append("&shortIntentUriOrDefault=").append(2.toShort().toInt())
                    .append("&intIntentUriOrDefault=").append(3)
                    .append("&longIntentUriOrDefault=").append(4L)
                    .append("&floatIntentUriOrDefault=").append(5.toFloat())
                    .append("&doubleIntentUriOrDefault=").append(6.toDouble())
                    .append("&booleanIntentUriOrDefault=").append(true)
                    .append("&stringIntentUriRequired=").append("stringIntentUriRequired")
                    .append("&stringIntentUriOptional=").append("stringIntentUriOptional")
                    .append("&stringIntentUriOrDefault=").append("stringIntentUriOrDefault")
                    .append("&byteUriIntentOrDefault=").append(11.toByte().toInt())
                    .append("&shortUriIntentOrDefault=").append(12.toShort().toInt())
                    .append("&intUriIntentOrDefault=").append(13)
                    .append("&longUriIntentOrDefault=").append(14L)
                    .append("&floatUriIntentOrDefault=").append(15.toFloat())
                    .append("&doubleUriIntentOrDefault=").append(16.toDouble())
                    .append("&booleanUriIntentOrDefault=").append(true)
                    .append("&stringUriIntentRequired=").append("stringUriIntentRequired")
                    .append("&stringUriIntentOptional=").append("stringUriIntentOptional")
                    .append("&stringUriIntentOrDefault=").append("stringUriIntentOrDefault")
                    .toString()

            val uri = Uri.parse("tools4a://tools4a.com/panpf/tools4a/uri/intent?$params")

            val starter = Intent(context, TestBothIntentUriActivity::class.java)
            starter.action = Intent.ACTION_VIEW
            starter.data = uri
            starter.putExtra("byteIntentUriOrDefault", (-1).toByte())
            starter.putExtra("shortIntentUriOrDefault", (-2).toShort())
            starter.putExtra("intIntentUriOrDefault", -3)
            starter.putExtra("longIntentUriOrDefault", -4L)
            starter.putExtra("floatIntentUriOrDefault", -5f)
            starter.putExtra("doubleIntentUriOrDefault", -6.0)
            starter.putExtra("booleanIntentUriOrDefault", true)
            starter.putExtra("stringIntentUriRequired", "stringIntentRequired")
            starter.putExtra("stringIntentUriOptional", "stringIntentOptional")
            starter.putExtra("stringIntentUriOrDefault", "stringIntentOrDefault")

            starter.putExtra("byteUriIntentOrDefault", (-11).toByte())
            starter.putExtra("shortUriIntentOrDefault", (-12).toShort())
            starter.putExtra("intUriIntentOrDefault", -13)
            starter.putExtra("longUriIntentOrDefault", -14L)
            starter.putExtra("floatUriIntentOrDefault", -15f)
            starter.putExtra("doubleUriIntentOrDefault", -16.0)
            starter.putExtra("booleanUriIntentOrDefault", true)
            starter.putExtra("stringUriIntentRequired", "stringUriIntentRequired")
            starter.putExtra("stringUriIntentOptional", "stringUriIntentOptional")
            starter.putExtra("stringUriIntentOrDefault", "stringUriIntentOrDefault")
            return starter
        }
    }

    fun checkParams() {
        val activity = this

        val byteIntentUriOrDefault = activity.readByteIntentUriArgOr("byteIntentUriOrDefault", 0.toByte())
        val shortIntentUriOrDefault = activity.readShortIntentUriArgOr("shortIntentUriOrDefault", 0.toShort())
        val intIntentUriOrDefault = activity.readIntIntentUriArgOr("intIntentUriOrDefault", 0)
        val longIntentUriOrDefault = activity.readLongIntentUriArgOr("longIntentUriOrDefault", 0L)
        val floatIntentUriOrDefault = activity.readFloatIntentUriArgOr("floatIntentUriOrDefault", 0.toFloat())
        val doubleIntentUriOrDefault = activity.readDoubleIntentUriArgOr("doubleIntentUriOrDefault", 0.toDouble())
        val booleanIntentUriOrDefault = activity.readBooleanIntentUriArgOr("booleanIntentUriOrDefault", false)
        val stringIntentUriRequired = activity.readStringIntentUriArgOrThrow("stringIntentUriRequired")
        val stringIntentUriOrDefault = activity.readStringIntentUriArgOr("stringIntentUriOrDefault", "default")
        val stringIntentUriOrDefaultErrKey = activity.readStringIntentUriArgOr("stringIntentUriOrDefaultErrKey", "stringIntentUriOrDefaultErrKey")
        val stringIntentUriOptional = activity.readStringIntentUriArgOrNull("stringIntentUriOptional")
        val stringIntentUriOptionalErrKey = activity.readStringIntentUriArgOrNull("stringIntentUriOptionalErrKey")

        val byteUriIntentOrDefault = activity.readByteUriIntentArgOr("byteUriIntentOrDefault", 0.toByte())
        val shortUriIntentOrDefault = activity.readShortUriIntentArgOr("shortUriIntentOrDefault", 0.toShort())
        val intUriIntentOrDefault = activity.readIntUriIntentArgOr("intUriIntentOrDefault", 0)
        val longUriIntentOrDefault = activity.readLongUriIntentArgOr("longUriIntentOrDefault", 0L)
        val floatUriIntentOrDefault = activity.readFloatUriIntentArgOr("floatUriIntentOrDefault", 0.toFloat())
        val doubleUriIntentOrDefault = activity.readDoubleUriIntentArgOr("doubleUriIntentOrDefault", 0.toDouble())
        val booleanUriIntentOrDefault = activity.readBooleanUriIntentArgOr("booleanUriIntentOrDefault", false)
        val stringUriIntentRequired = activity.readStringUriIntentArgOrThrow("stringUriIntentRequired")
        val stringUriIntentOrDefault = activity.readStringUriIntentArgOr("stringUriIntentOrDefault", "default")
        val stringUriIntentOrDefaultErrKey = activity.readStringUriIntentArgOr("stringUriIntentOrDefaultErrKey", "stringUriIntentOrDefaultErrKey")
        val stringUriIntentOptional = activity.readStringUriIntentArgOrNull("stringUriIntentOptional")
        val stringUriIntentOptionalErrKey = activity.readStringUriIntentArgOrNull("stringUriIntentOptionalErrKey")

        Assert.assertTrue(byteIntentUriOrDefault == (-1).toByte())
        Assert.assertTrue(shortIntentUriOrDefault == (-2).toShort())
        Assert.assertTrue(intIntentUriOrDefault == -3)
        Assert.assertTrue(longIntentUriOrDefault == -4L)
        Assert.assertTrue(floatIntentUriOrDefault == -5f)
        Assert.assertTrue(doubleIntentUriOrDefault == -6.0)
        Assert.assertTrue(booleanIntentUriOrDefault)
        Assert.assertTrue(stringIntentUriRequired == "stringIntentRequired")
        Assert.assertTrue(stringIntentUriOptional == "stringIntentOptional")
        Assert.assertNull(stringIntentUriOptionalErrKey)
        Assert.assertTrue(stringIntentUriOrDefault == "stringIntentOrDefault")
        Assert.assertTrue(stringIntentUriOrDefaultErrKey == "stringIntentUriOrDefaultErrKey")

        Assert.assertTrue(byteUriIntentOrDefault == 11.toByte())
        Assert.assertTrue(shortUriIntentOrDefault == 12.toShort())
        Assert.assertTrue(intUriIntentOrDefault == 13)
        Assert.assertTrue(longUriIntentOrDefault == 14L)
        Assert.assertTrue(floatUriIntentOrDefault == 15f)
        Assert.assertTrue(doubleUriIntentOrDefault == 16.0)
        Assert.assertTrue(booleanUriIntentOrDefault)
        Assert.assertTrue(stringUriIntentRequired == "stringUriIntentRequired")
        Assert.assertTrue(stringUriIntentOptional == "stringUriIntentOptional")
        Assert.assertNull(stringUriIntentOptionalErrKey)
        Assert.assertTrue(stringUriIntentOrDefault == "stringUriIntentOrDefault")
        Assert.assertTrue(stringUriIntentOrDefaultErrKey == "stringUriIntentOrDefaultErrKey")
    }
}